plugin/src/main/java/io/fabric8/maven/plugin/BuildMojo.java

```diff
@@ -241,7 +293,191 @@ private void executeOpenShiftBuild(ServiceHub hub, ImageConfiguration imageConfi
+
+        waitForOpenShiftBuildToComplete(client, buildName, build);
+
+        updateImageStreamTags(client, imageConfig, buildName, build);
+    }
+
+
+    private void waitForOpenShiftBuildToComplete(OpenShiftClient client, String buildConfigName, Build build) {
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        String buildName = getName(build);
+        Watcher<Build> buildWatcher = new Watcher<Build>() {
+            @Override
+            public void eventReceived(Action action, Build resource) {
+                if (isBuildCompleted(action, resource)) {
+                    latch.countDown();
+                }
+            }
+
+            @Override
+            public void onClose(KubernetesClientException cause) {
+            }
+        };
+        log.info("Waiting for build " + buildName + " to complete...");
+        try (Watch watcher = client.builds().withName(buildName).watch(buildWatcher)) {
+            while (latch.getCount() > 0L) {
+                try {
+                    latch.await();
+                } catch (InterruptedException e) {
+                    // ignore
+                }
+            }
+            log.info("Build " + buildName + " completed!");
+        }
+    }
+
+
+    /**
+     * Lets update the ImageStream to include the correct tags
+     */
+    private void updateImageStreamTags(OpenShiftClient client, ImageConfiguration imageConfig, String buildConfigName, Build build) throws MojoExecutionException {
+
+        try {
+            File manifest = openshiftManifest;
+            if (!Files.isFile(manifest)) {
+                throw new MojoFailureException("No such generated manifest file: " + manifest);
+            }
+
+            String namespace = clusterAccess.getNamespace();
+            Controller controller = new Controller(client);
+
+            boolean updated = false;
+            Set<HasMetadata> entities = loadResources(client, controller, namespace, manifest, project, log);
+            for (HasMetadata entity : entities) {
+                if (entity instanceof ImageStream) {
+                    ImageStream is = (ImageStream) entity;
+                    String imageStreamName = KubernetesHelper.getName(is);
+                    if (Objects.equals(buildConfigName, imageStreamName)) {
+                        if (updateImageStreamTag(client, imageConfig, is, buildConfigName)) {
+                            updated = true;
+                        }
+                    }
+                }
+            }
+            if (updated) {
+                // lets store the entities again!
+                KubernetesList entity = new KubernetesListBuilder().withItems(new ArrayList<>(entities)).build();
+                File resourceFileBase = new File(this.targetDir, ResourceClassifier.OPENSHIFT.getValue());
+                AbstractResourceMojo.writeResourcesIndividualAndComposite(entity, resourceFileBase, ResourceFileType.yaml, log);
```

Comment:
We need to distribute the descriptors as part of the jar. The issue is that the `ResourceMojo` doesn't generate complete resources; its a combination of both Mojos which does. Just like `process-resources` and `compile` make the jar contents in typical java apps. 

Maybe its easier to digest if the `BuildMojo` is the mojo which appends the `ImageStream` resources? (If it were not for the combining of resources into uber-yamls, then the `BuildMojo` would just be generating its own resources in addition to those created by `ResourceMojo` - then ordering is less important really too. To simplify things I'm tempted to refactor the logic along these lines; that `ResourceMojo` generates most resources; then `BuildMojo` appends `ImageStreams` in openshift mode; then they are logically combined - I'd also like to make them order agnostic (so re-running `ResourceMojo` in isolation won't zap all the `ImageStream` resources generated from the previous `BuildMojo` goal).

On point 2; the ImageTag is how to install the image stream into other namespaces for single cluster openshift environments. e.g. if folks have 1 cluster, they build/release in one namespace then reuse those same resources to install in any other namespace. I'm hoping we can figure out how to also make this ImageTag multi-cluster aware (I'm hoping its just a case of also including the URL of the internal openshift docker registry too) - in which case they will be installation agnostic (provided the cluster that built the image stays up). Not 100% sure as of yet on the multi-cluster nature - but its my hope we can make them multi-cluster-aware somehow. Otherwise its time to drop ImageStreams completely as not fit for purpose and switch tacks and use regular docker registries again ;). 

To say that another way; I don't think open source projects would never use OpenShift S2I builders or ImageStreams; we'd just use docker hub instead and use real docker registries. But when developing on premise releases using OpenShift and S2I then you do want to put the ImageTag in your releases so that you can use `oc apply` to install the released resources & image in any namespace